proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,538 @@
+    return {Point, Rect};
+}
+
+const {Point, Rect} = makeTypes();
+
+WebAssembly.instantiateStreaming(fetch('example4.wasm'), {'':{Point}})
+.then(({instance}) => {
+    const r = new Rect(new Point(1,2), new Point(3,4));
+    instance.exports.goWild(r);  // succeeds
+    assert(r.x1 instanceof Point);
+    assert(r.x1.x === 10);
+    assert(r.x1.y === 20);
+
+    const {Point2:Point, Rect2:Rect} = makeTypes();
+    const r2 = new Rect2(new Point2(1,2), new Point2(3,4));
+    instance.exports.goWild(r2);  // throws at call boundary
+});
+```
+
+Thus, the expected nominal invariants of JS can be preserved by a combination of
+instantiation-time and run-time checks.
+
+But what about the other direction: what if WebAssembly exports a struct
+type definition that uses a `ref` field type? To reflect WebAssembly structural
+subtyping, we need new Value Type objects. Since the set of Value Type objects
+is open-ended, the WebAssembly JS API adds a new Value Type constructor
+function: `WebAssembly.ref(T)`. Using this, JS can create a suitable Type
+Definition object whose instances can be passed to the above `example3.wasm`
+because its fields are structurally typed:
+
+```js
+function makeTypes() {
+    const Point = new StructType([{name: "x", type: int32}, {name: "y", type: int32}]);
+    const PointRef = WebAssembly.ref(Point);
+    const Rect = new StructType([{name: "p1", type: PointRef}, {name: "p2", type: PointRef}]);
+    return {Point, Rect};
+}
+
+WebAssembly.instantiateStreaming(fetch('example3.wasm'))
+.then(({instance}) => {
+    const {Point, Rect} = makeTypes();
+    var rect = new Rect(new Point(1,2), new Point(3,4));
+    assert("x" in rect.x1);
+    instance.exports.goWild(rect);
+    assert(!("x" in rect.x1));
+    assert(rect.x1[0] === 10);
+    assert(rect.x1[1] === 20);
+
+    const {Point2:Point, Rect2:Rect} = makeTypes();
+    var rect2 = new Rect2(new Point2(1,2), new Point2(3,4));
+    assert(rect.__proto__ !== rect2.__proto__);
+    instance.exports.goWild(rect2);  // ok: structurally equivalent
+});
+```
+
+Having a distinction between these two kinds of subtyping is unfortunate, but
+the expectation is that practical uses will generate all these Type Definition
+objects as part of the usual JS boilerplate glue code and end users would simply
+interact with objects that were either tuple-y or had property names as expected
+from the source language. (For example, a WebAssembly-friendly [Flow](https://flow.org)
+variant might use structural typing for fields with [Tuple Type](https://flow.org/en/docs/types/tuples/)
+and nominal typing for fields with [Class Type](https://flow.org/en/docs/types/classes/).)
+
+In addition to the dynamic checks necessary when JS passes a value to
+WebAssembly (either as an argument to a wasm exported function or a return value
+from a wasm imported function), the instantiation algorithm must ensure that
```

Comment:
Or through a global variable or via a slot in a table with a nontrivial ref type?

Reply:
Good point; resolving by adding "etc" linked down to the last section which specifically calls out these two cases.